### PR TITLE
asyncify promisified functions

### PR DESCRIPTION
**Pull request by zartdinov**
Most of Promise realisations doesn't contain `nodeify` method. Here is extention of async wrapper for promisified functions.
For example promise-styled [openpgp](https://github.com/openpgpjs/openpgpjs) library functions looks like
```javascript
openpgp.key.generate(options).then(function(key) {
    // success
}).catch(function(error) {
    // failure
});
```
And can be easily wrapped to use with favourite async:)
```javascript
async.asyncify(openpgp.key.generate);
```

### CODE REVIEW

File: lib/async.js [hunk 1]
```diff
@@ -1186,10 +1186,18 @@
             var result;
             try {
                 result = func.apply(this, args);
+                // if result is Promise object
+                if (typeof result.then === "function") {
+                    result.then(function(values) {
+                        var args = [null].concat(values);
+                        callback.apply(this, args);
```
Comment (megawac) [inline lib/async.js]: There can be only one value from promise so can you do 
```js
callback.call(null, value)
```

**Comment by megawac** [comment]
:+1: can you add some unit tests (feel free to add a dev dependency for some promise implementation, e.g. NPO/rsvp/bluebird/es6-promise)

File: lib/async.js [hunk 2]
```diff
@@ -1186,10 +1186,18 @@
...
+                    }).catch(callback);
+                } else {
+                    callback(null, result);
```
Comment (aearly) [inline lib/async.js]: We don't want to have the callback call in the try/catch block.

Comment (aearly) [inline lib/async.js]: Yep: The failing test is the one that guards against this: https://travis-ci.org/caolan/async/jobs/70115957

**Comment by zartdinov** [comment]
Done!

File: test/test-async.js [hunk 3]
```diff
@@ -4267,5 +4267,49 @@ exports['asyncify'] = {
...
-    }
+    },
+
+    'promisified': [
+        'native-promise-only',
+        'bluebird',
+        'es6-promise',
+        'rsvp'
+    ].reduce(function(promises, name) {
+        var Promise = require(name);
```
Comment (aearly) [inline test/test-async.js]: You might want to make these tests node-only, because this won't work in the browser.

**Comment by zartdinov** [comment]
I think node-only for a while, but all promise libraries differently support browsers.

**Comment by aearly** [comment]
It's not the promise implementation, but the `require(name)` statement that won't work.  We can fix the tests for browsers later.

**Pull request merged**